src/System.Private.CoreLib/shared/System/InvariantArray.cs

```diff
@@ -0,0 +1,181 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Runtime.CompilerServices;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    /// <summary>
+    /// A wrapper around an "invariant" <typeparamref name="T"/> array; that is, without
+    /// any covariance or contravariance.
+    /// </summary>
+    internal readonly struct InvariantArray<T>
+    {
+        private readonly T[] _array;
+
+        /// <summary>
+        /// Creates an invariant wrapper around an existing array.
+        /// <paramref name="array"/> must be null or exactly T[] (no covariance or contravariance).
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public InvariantArray(T[] array)
+        {
+            if (RuntimeHelpers.IsReference<T>() && (array != null && array.GetType() != typeof(T[])))
+            {
+                ThrowHelper.ThrowArrayTypeMismatchException();
+            }
+
+            _array = array;
+        }
+
+        /// <summary>
+        /// Allocates a new array of the specified length and creates an invariant
+        /// wrapper around it.
+        /// </summary>
+        /// <param name="count"></param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public InvariantArray(int count)
+        {
+            _array = new T[count];
+        }
+
+        /// <summary>
+        /// Provides access to the backing array.
+        /// </summary>
+        public T[] Array
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _array;
+        }
+
+        /// <summary>
+        /// Returns true if the backing array is null.
+        /// </summary>
+        public bool IsNull
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _array is null;
+        }
+
+        /// <summary>
+        /// Returns the length of the backing array.
+        /// </summary>
+        /// <exception cref="NullReferenceException">
+        /// If <see cref="IsNull"/> is true.
+        /// </exception>
+        public int Length
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            [StackTraceHidden]
+            get => _array.Length;
+        }
+
+        /// <summary>
+        /// Returns the backing array as a mutable <see cref="Span{T}"/>; or
+        /// <see cref="Span{T}.Empty"/> if <see cref="IsNull"/> is true.
+        /// </summary>
+        public Span<T> Span
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (IsNull)
+                {
+                    return default;
+                }
+                else
+                {
+                    return new Span<T>(ref Unsafe.As<byte, T>(ref _array.GetRawSzArrayData()), _array.Length);
```

Comment:
Write in the same way as in L112? (i.e. "Accessing Length first avoids an unnecessary cmp instruction.")

Reply:
We've already incurred the null check at the top of the property getter (note `IsNull` access), so may as well write this the natural way.

Reply:
Ah, I see the JIT doesn't reload `_array` in `else`. Thx.